### ISSUE TITLE
Error: spawn ../node_modules/optipng-bin/vendor/optipng ENOENT

### ISSUE THREAD

**Issue by owcc** [imagemin/optipng-bin#84]
I found that **../node_modules/optipng-bin/vendor/optipng** is not found caused this problem, and I run `node lib/install.js` resolve this issues, is not should be auto-run after `npm install optipng-bin`?

so happens with **jpegtran-bin** too.
(sorry for my pool english)

**Comment by maxheadroom** [comment]
Same problem here. the ```node lib/install.js``` but doesn't solve the problem:

```
/build/node_modules/optipng-bin # node lib/install.js
  ⚠ spawn /build/node_modules/optipng-bin/vendor/optipng ENOENT
  ⚠ optipng pre-build test failed
  ℹ compiling from source
  ✖ Error: ./configure --with-system-zlib --prefix="/build/node_modules/optipng-bin/vendor" --bindir="/build/node_modules/optipng-bin/vendor" && make install
Command failed: ./configure --with-system-zlib --prefix="/build/node_modules/optipng-bin/vendor" --bindir="/build/node_modules/optipng-bin/vendor"

    at ChildProcess.exithandler (child_process.js:206:12)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:877:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
```

**Comment by Harrisonbro** [comment]
Yep, same problem for me:

```
[~/public_html/envoyer/releases/20170227151550]# npm install optipng-bin

> optipng-bin@3.1.2 postinstall /home/public_html/envoyer/releases/20170227151550/node_modules/optipng-bin
> node lib/install.js

  ⚠ spawn /home/public_html/envoyer/releases/20170227151550/node_modules/optipng-bin/vendor/optipng EAGAIN
  ⚠ optipng pre-build test failed
  ℹ compiling from source
  ✖ Error: ./configure --with-system-zlib --prefix="/home/public_html/envoyer/releases/20170227151550/node_modules/optipng-bin/vendor" --bindir="/home/public_html/envoyer/releases/20170227151550/node_modules/optipng-bin/vendor" && make install
spawn /bin/sh EAGAIN
    at exports._errnoException (util.js:1026:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:193:32)
    at onErrorNT (internal/child_process.js:359:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
countryside-classroom-2016@1.0.0 /home/public_html/envoyer/releases/20170227151550
└─┬ gulp-imagemin@2.4.0
  └─┬ imagemin@4.0.0
    └─┬ imagemin-optipng@4.3.0
      └── optipng-bin@3.1.2 
```

**Comment by Harrisonbro** [comment]
It seems to install fine on Node v5.6.0 (npm v3.6.0) but fails with the above output on Node 6.

**Comment by 1000ch** [comment]
This bay be fixed on #86, would you try again?

**Comment by Harrisonbro** [comment]
Yes, it appears to no longer be an issue. I can't replicate on Node 5, 6 or 7.

**Comment by mblarsen** [comment]
Happens for me on Node 8.5

**Comment by twhetzel** [comment]
Happens for me on Node 8.6

**Comment by ladotonia** [comment]
I had the same issue with Node 7.10.1 on an alpine docker. I was missing zlib-dev and install.js runs configure using --with-system-zlib. The err.stack output doesn't show why it failed.

**Comment by LazyChenY** [comment]
`npm rebuild`
works for me.

**Comment by brabeji** [comment]
I've just had very funny debugging session:
```
> optipng-bin@3.1.4 postinstall /opt/app/node_modules/optipng-bin
> node lib/install.js

  ⚠ spawn /opt/app/node_modules/optipng-bin/vendor/optipng ENOENT
  ⚠ optipng pre-build test failed
  ℹ compiling from source
  ✖ Error: ./configure --with-system-zlib --prefix="/opt/app/node_modules/optipng-bin/vendor" --bindir="/opt/app/node_modules/optipng-bin/vendor" && make install
Command failed: ./configure --with-system-zlib --prefix="/opt/app/node_modules/optipng-bin/vendor" --bindir="/opt/app/node_modules/optipng-bin/vendor"
/bin/sh: ./configure: not found

    at ChildProcess.exithandler (child_process.js:206:12)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:877:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
```
*TL;DR* SourceForge is down

## Course of failure
1. Prebuilt binary is not working on my platform. No worries, let `optipng-bin` build it from sources.
2. `optipng-bin` uses `bin-build` to download gzipped archive of `optipng` source from SourceForge and run build commands i.e. `./configure --with-system-zlib ...`
3. SourceForge in disaster recovery mode responds with temporary redirect to html document. This document contains javascript redirect to backup archive file on amazon. The html is thus saved as the contents of ZIP archive.
4. `bin-build` tries to unzip garbage, fails and says nothing about it. Runs build commands in an empty directory, resulting in the above error.

I loved it.

**Comment by AndrewSavetchuk** [comment]
@LazyChenY thanks!

**Comment by timmliu** [comment]
removing entire node_modules folder didn't work for me.. but what worked was to delete optipng-bin folder inside of /node_modules, run my build script again, and then it may complain about another folder inside of /node_modules, so delete that too, then run build script again, etc repeat. there were 4-5 modules i had to manually delete but then the problem went away after that. for reference i'm on node v9.11.2 and npm 5.6.0

**Comment by damianobarbati** [comment]
Same problem on alpine using yarn (trying to run webpack when building a docker image). 

Did anyone had luck with that?

**Comment by Renrhaf** [comment]
Hi, I was using a Docker Linux Alpine image and missing some extensions causing this error to happen. Try installing the following packaging in your Dockerfile to fix it : 
`pkgconfig autoconf automake libtool nasm`

**Comment by twoheaded** [comment]
Thanks @Renrhaf. I had to install `build-base` and `zlib-dev` else in my Alpine Docker container:

```
RUN apk --no-cache add pkgconfig autoconf automake libtool nasm build-base zlib-dev
```

**Comment by bugwheels94** [comment]
It is pkg-config

**Comment by gvlasov** [comment]
Thanks @twoheaded. I had to install ``libc6-compat` and `libjpeg-turbo-dev` in my Alpine Docker container:

    RUN apk --no-cache add pkgconfig autoconf automake libtool nasm build-base zlib-dev libc6-compat libjpeg-turbo-dev

**Comment by 137px** [comment]
I also had to manually remove the optipng-bin folder after npm install and then run npm install again to get it to work

**Comment by tavurth** [comment]
Since this problem happens much earlier in the build than the final error which breaks the process, perhaps it would be nicer to add a fairly large warning text when this error occurs?

```
********************************************************************************
********************************************************************************
                         _____                         
                        | ____| _ __  _ __  ___   _ __ 
                        |  _|  | '__|| '__|/ _ \ | '__|
                        | |___ | |   | |  | (_) || |   
                        |_____||_|   |_|   \___/ |_|   
                                                       
                             Source forge is down
                             Optipng will not build

********************************************************************************
********************************************************************************
```

**Comment by neilveil** [comment]
### It's a version issue

There are two cases:

I)  It the project is working fine on other systems
II) It you only have initialized the project & it's showing this error

Solution:

- Match node version first if case I is applicable, you can use `nvm` for that
- `cd`
- `sudo chmod 755 -R .npm`
- `cd /you/project/dir`
- `sudo rm -rf node_modules`
- `npm rebuid`
- `npm i`

**Your welcome!**

**Comment by DiegoMarcelino1992** [comment]
i'm running yarn instal on pm2 server 
ERROR =>>>

[4/4] Building fresh packages...
[1/10] ⠂ gifsicle
[7/10] ⠂ cwebp-bin
[-/10] ⠂ waiting...
[4/10] ⠂ mozjpeg
error /var/www/node/next/node_modules/optipng-bin: Command failed.
Exit code: 1
Command: node lib/install.js
Arguments: 
Directory: /var/www/node/next/node_modules/optipng-bin
Output:
⚠ Command failed: /var/www/node/next/node_modules/optipng-bin/vendor/optipng --version
/var/www/node/next/node_modules/optipng-bin/vendor/optipng: 1: /var/www/node/next/node_modules/optipng-bin/vendor/optipng: @@8�@@@@@@@��@@@@����: not found
/var/www/node/next/node_modules/optipng-bin/vendor/optipng: 2: /var/www/node/next/node_modules/optipng-bin/vendor/optipng: d: not found
/var/www/node/next/node_modules/optipng-bin/vendor/optipng: 1: /var/www/node/next/node_modules/optipng-bin/vendor/optipng: Syntax error: ";" unexpected
/var/www/node/next/node_modules/optipng-bin/vendor/optipng: 1: /var/www/node/next/node_modules/optipng-bin/vendor/optipng: ELF: not found


  ⚠ optipng pre-build test failed
  ℹ compiling from source
  ✖ Error: Command failed: /bin/sh -c make install
pngrtran.c:99:1: warning: ‘png_rtran_ok’ defined but not used [-Wunused-function]
 png_rtran_ok(png_structrp png_ptr, int need_IHDR)
 ^
ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
pngxmem.c: In function ‘pngx_malloc_rows_extended’:
pngxmem.c:38:34: warning: comparison is always false due to limited range of data type [-Wtype-limits]
        (pngx_alloc_size_t)height > (pngx_alloc_size_t)(-1) / sizeof(png_bytep))
                                  ^
ar: `u' modifier ignored since `D' is the default (see `U')
../libpng/libpng.a(pngrutil.o): In function `png_read_filter_row':
pngrutil.c:(.text+0x29ec): undefined reference to `png_init_filter_functions_vsx'
collect2: error: ld returned 1 exit status
make[1]: *** [optipng] Error 1
make: *** [install] Error 2

cd src/optipng && \
make install && \
cd ../..
make[1]: Entering directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/optipng'
cd ../libpng && \
make -f Makefile PNGLIBCONF_H_PREBUILT=pnglibconf.h.optipng && \
cd ../optipng
make[2]: Entering directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/libpng'
cp pnglibconf.h.optipng pnglibconf.h
gcc -c -I../zlib  -O2 -Wall -Wextra -o png.o png.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngerror.o pngerror.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngget.o pngget.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngmem.o pngmem.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngpread.o pngpread.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngread.o pngread.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngrio.o pngrio.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngrtran.o pngrtran.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngrutil.o pngrutil.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngset.o pngset.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngtrans.o pngtrans.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwio.o pngwio.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwrite.o pngwrite.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwtran.o pngwtran.c
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwutil.o pngwutil.c
ar rcs libpng.a png.o pngerror.o pngget.o pngmem.o pngpread.o pngread.o pngrio.o pngrtran.o pngrutil.o pngset.o pngtrans.o pngwio.o pngwrite.o pngwtran.o pngwutil.o
ranlib libpng.a
gcc -c -I../zlib  -O2 -Wall -Wextra -o pngtest.o pngtest.c
gcc  -L../zlib -o pngtest pngtest.o libpng.a -lz -lm
make[2]: Leaving directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/libpng'
cd ../opngreduc && \
make -f Makefile libopngreduc.a && \
cd ../optipng
make[2]: Entering directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/opngreduc'
gcc -c  -O2 -Wall -Wextra  -I../libpng -o opngreduc.o opngreduc.c
ar cru libopngreduc.a opngreduc.o
ranlib libopngreduc.a
make[2]: Leaving directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/opngreduc'
cd ../gifread && \
make -f Makefile libgifread.a && \
cd ../optipng
make[2]: Entering directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/gifread'
gcc -c  -O2 -Wall -Wextra -o gifread.o gifread.c
ar cru libgifread.a gifread.o
ranlib libgifread.a
make[2]: Leaving directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/gifread'
cd ../pnmio && \
make -f Makefile libpnmio.a && \
cd ../optipng
make[2]: Entering directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/pnmio'
gcc -c  -O2 -Wall -Wextra -o pnmin.o pnmin.c
gcc -c  -O2 -Wall -Wextra -o pnmout.o pnmout.c
gcc -c  -O2 -Wall -Wextra -o pnmutil.o pnmutil.c
ar cru libpnmio.a pnmin.o pnmout.o pnmutil.o
ranlib libpnmio.a
make[2]: Leaving directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/pnmio'
cd ../minitiff && \
make -f Makefile libminitiff.a && \
cd ../optipng
make[2]: Entering directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/minitiff'
gcc -c  -O2 -Wall -Wextra -o tiffread.o tiffread.c
gcc -c  -O2 -Wall -Wextra -o tiffutil.o tiffutil.c
ar cru libminitiff.a tiffread.o tiffutil.o 
ranlib libminitiff.a
make[2]: Leaving directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/minitiff'
cd ../pngxtern && \
make -f Makefile libpngxtern.a && \
cd ../optipng
make[2]: Entering directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/pngxtern'
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxread.o pngxread.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrbmp.o pngxrbmp.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrgif.o pngxrgif.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrjpg.o pngxrjpg.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrpnm.o pngxrpnm.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrtif.o pngxrtif.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxio.o pngxio.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxmem.o pngxmem.c
gcc -c  -O2 -Wall -Wextra  -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxset.o pngxset.c
ar cru libpngxtern.a pngxread.o pngxrbmp.o pngxrgif.o pngxrjpg.o pngxrpnm.o pngxrtif.o pngxio.o pngxmem.o pngxset.o
ranlib libpngxtern.a
make[2]: Leaving directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/pngxtern'
gcc -c  -O2 -Wall -Wextra -I../cexcept  -I../libpng -I../opngreduc -I../pngxtern -o optipng.o optipng.c
gcc -c  -O2 -Wall -Wextra -I../cexcept  -I../libpng -I../opngreduc -I../pngxtern -o optim.o optim.c
gcc -c  -O2 -Wall -Wextra -I../cexcept  -I../libpng -I../opngreduc -I../pngxtern -o bitset.o bitset.c
gcc -c  -O2 -Wall -Wextra -I../cexcept  -I../libpng -I../opngreduc -I../pngxtern -o ioutil.o ioutil.c
gcc -c  -O2 -Wall -Wextra -I../cexcept  -I../libpng -I../opngreduc -I../pngxtern -o ratio.o ratio.c
gcc -c  -O2 -Wall -Wextra -I../cexcept  -I../libpng -I../opngreduc -I../pngxtern -o wildargs.o wildargs.c
gcc -s -o optipng optipng.o optim.o bitset.o ioutil.o ratio.o wildargs.o ../opngreduc/libopngreduc.a ../pngxtern/libpngxtern.a ../libpng/libpng.a  ../gifread/libgifread.a ../pnmio/libpnmio.a ../minitiff/libminitiff.a  -lz -lm 
Makefile:100: recipe for target 'optipng' failed
make[1]: Leaving directory '/tmp/eed0390b-499e-47ea-8fc9-768b0ead27c4/src/optipng'
Makefile:14: recipe for target 'install' failed

**Comment by gangsthub** [comment]
In my case was a combination of 2 factors, solved by:

```
npm i -g node-gyp
npm rebuild
```

**Comment by ComahonQQComahon** [comment]
Hi!
pls help me!



`C:\Users\Сергей\Desktop\myapp>npm i gulp-image

> gifsicle@4.0.1 postinstall C:\Users\Сергей\Desktop\myapp\node_modules\gifsicle

> node lib/install.js

  √ gifsicle pre-build test passed successfully

> guetzli@2.0.0 postinstall C:\Users\Сергей\Desktop\myapp\node_modules\guetzli
> node lib/install.js

  ‼ Command failed: C:\Users\Сергей\Desktop\myapp\node_modules\guetzli\vendor\gu
etzli.exe --verbose
Guetzli JPEG compressor. Usage:
guetzli [flags] input_filename output_filename

Flags:
  --verbose    - Print a verbose trace of all attempts to standard output.
  --quality Q  - Visual quality to aim for, expressed as a JPEG quality value.
  --memlimit M - Memory limit in MB. Guetzli will fail if unable to stay under
                 the limit. Default is 6000 MB
  --nomemlimit - Do not limit memory usage.


  ‼ guetzli pre-build test failed
  i compiling from source
  × Error: Command failed: C:\Windows\system32\cmd.exe /s /c "mkdir -p C:\Users\
Сергей\Desktop\myapp\node_modules\guetzli\vendor"
�������� ��� 䠩� C:\Users\��ࣥ�\Desktop\myapp\node_modules\guetzli\vendor 㦥 ������
�.
�訡�� �� �६� ��ࠡ�⪨: C:\Users\��ࣥ�\Desktop\myapp\node_modules\guetzli\vendor.


    at C:\Users\Сергей\Desktop\myapp\node_modules\execa\index.js:231:11
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 0)

> jpeg-recompress-bin@4.0.0 postinstall C:\Users\Сергей\Desktop\myapp\node_modul
es\jpeg-recompress-bin
> node lib/install.js

  × Error: spawn UNKNOWN
    at ChildProcess.spawn (internal/child_process.js:405:11)
    at Object.spawn (child_process.js:548:9)
    at module.exports (C:\Users\Сергей\Desktop\myapp\node_modules\execa\index.js
:139:26)
    at C:\Users\Сергей\Desktop\myapp\node_modules\bin-check\index.js:16:11

> mozjpeg@6.0.1 postinstall C:\Users\Сергей\Desktop\myapp\node_modules\mozjpeg
> node lib/install.js

  √ mozjpeg pre-build test passed successfully

> optipng-bin@6.0.0 postinstall C:\Users\Сергей\Desktop\myapp\node_modules\optip
ng-bin
> node lib/install.js

  ‼ spawn UNKNOWN
  ‼ optipng pre-build test failed
  i compiling from source
  × Error: Command failed: C:\Windows\system32\cmd.exe /s /c "./configure --with
-system-zlib --prefix="C:\Users\Сергей\Desktop\myapp\node_modules\optipng-bin\ve
ndor" --bindir="C:\Users\Сергей\Desktop\myapp\node_modules\optipng-bin\vendor""
"." �� ���� ����७��� ��� ���譥�
��������, �ᯮ��塞�� �ணࠬ��� ��� ������ 䠩���.


    at C:\Users\Сергей\Desktop\myapp\node_modules\execa\index.js:231:11
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 0)
npm WARN myapp@1.0.0 No description
npm WARN myapp@1.0.0 No repository field.
npm WARN myapp@1.0.0 No license field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! optipng-bin@6.0.0 postinstall: `node lib/install.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the optipng-bin@6.0.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional log
ging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Сергей\AppData\Roaming\npm-cache\_logs\2020-01-22T04_32_20
_168Z-debug.log

C:\Users\Сергей\Desktop\myapp>`

Tell me the solution to this problem! thanks

**Comment by harmenjanssen** [comment]
@twoheaded Thanks! I was also able to fix my Alpine container by switching to `zlib-dev` instead of `zlib`.

**Comment by experian-greg-myers** [comment]
For me it was corporate firewall was MITM the connection causing build failure, fixed by telling NODE to use system CA:
```
rm -rf node_modules

export NODE_OPTIONS=--use-openssl-ca

yarn install

bin/webpacker
```

**Comment by idevmans** [comment]
> `npm rebuild`
> works for me.

Kinda weirdooo, cuz i install every single dependencies, anyway it works to me 👍 
WSL2 Win 20h2 
node 15.4.0 
npm  7.0.15

**Comment by skilletComatose** [comment]
1) npm install --save-dev imagemin imagemin-cli
2) npm rebuild 
solved the problem to me

**Comment by yuweiwen007** [comment]
ENOENT mean no entry file, so touch entry file👇👇👇

1. node node_modules/optipng-bin/lib/install.js
2. tar xzf node_modules/optipng-bin/vendor/source/optipng.tar.gz

**Comment by artemijeka** [comment]
`node node_modules/optipng-bin/lib/install.js`

**Comment by maytan-we** [comment]
> Thanks @Renrhaf. I had to install `build-base` and `zlib-dev` else in my Alpine Docker container:
> 
> ```
> RUN apk --no-cache add pkgconfig autoconf automake libtool nasm build-base zlib-dev
> ```

This solved all my node problems. Thanks @twoheaded

**Comment by waybackarchiver** [comment]
> Thanks @Renrhaf. I had to install `build-base` and `zlib-dev` else in my Alpine Docker container:
> 
> ```
> RUN apk --no-cache add pkgconfig autoconf automake libtool nasm build-base zlib-dev
> ```

Thanks you. This worked for me, but with nsam, not [yasm](https://yasm.tortall.net/).

**Comment by HokageKonohi** [comment]
Что делать если node_modules\optipng-bin\vendor"
